### PR TITLE
Provide consistent exception handling.

### PR DESCRIPTION
**Pull request by RackerWilliams**
All exceptions provided by attribute mapper will now be derived from
`AttributeMapperException`.

Also provide a new JSON Schema to validate JSON/YAML policies in order
to get meaningful error messages.

The following are the description of new exceptions:

AttributeMapperException extends Exception:

Base class for all other exceptions. This is also used to wrap
exceptions we don't understand. In this case, the message will be
generic : "Error while processing mapping policy".  This may be an
internal issue with the attribute mapper.

KnownAttributeMapperException extends AttributeMapperException:

Base class for all "known" exceptions.  In this case the message is
specific to the exception and may be sent to the end user.

ParseException extends KnownAttributeMapperException:

Malformed XML, JSON, YAML.

ValidationException extends KnownAttributeMapperException:

There's a validation error -- something which validates XSD or
JsonSchema.

XPathException extends ValidationException

There's an issue with an XPath in the policy.

### CODE REVIEW

**Comment by dmnjohns** [comment]
Yup, that's the same behavior as not catching a `Throwable` at all.

**Comment by dmnjohns** [comment]
In the `AttributeMapper` class, should `parseJsonNode` and `parseYamlNode` also handle exceptions? They are publicly exposed methods, so for the sake of consistency, and so that users don't have to directly depend on Jackson, I lean toward doing so.

**Comment by RackerWilliams** [comment]
I actually regret making those calls public.  They are a convenience function, always made within the contexts of other functions.  Will I break repose if I make them private?

File: core/src/test/scala/com/rackspace/identity/components/XPathAssertions.scala [hunk 1]
```diff
@@ -87,7 +86,7 @@ trait XPathAssertions {
         throw new TestFailedException (s"XPath expression does not evaluate to true(): $xpathString", 4) // scalastyle:ignore
       }
     } catch {
-      case xpe : XPathException => throw new TestFailedException (s"Error in XPath $xpathString", xpe, 4) // scalastyle:ignore
+      case xpe : javax.xml.xpath.XPathException => throw new TestFailedException (s"Error in XPath $xpathString", xpe, 4) // scalastyle:ignore
```
[the remark this answers is not in the document]
Comment (RackerWilliams) [reply core/src/test/scala/com/rackspace/identity/components/XPathAssertions.scala]: Yea, it probably is.  I'm really not sure what to name the magic number actually :-)

**Comment by dmnjohns** [comment]
Heh, so the reason I brought those functions up is actually that Repose is using them in the policy validation filter. It gives us an easy way to take a YAML or JSON policy from the request body, translate that to a data structure that the `validatePolicy` function understands, and pass the result to the `validatePolicy` method. That is, we certainly appreciate those being convenience functions. We could, of course, use Jackson directly... but personally, I would prefer the translation stay there since the Attribute Mapper library is already dealing with those conversions to the right data structures. That would allow the policy validation filter to be a thin wrapper around the `validatePolicy` functions in Attribute Mapper.

Of course, I totally understand not wanting those functions to be public. This is not, after all, a data format conversion tool. So perhaps instead we can expose a variant of `validatePolicy` to accept either a YAML, JSON, or XML argument? That variant can, in turn, make use of the private conversion functions. And, of course, Repose gets to keep its hands clean of those conversions. I never really loved the `validatePolicy` function requiring a `JsonNode` anyway since that requires the user of the library to go from whatever generic data they may have to a library-specific structure (thus constraining the usage of the library).

**tl;dr: ** Making those functions private will break Repose, and while we could fix that, I still don't love the idea of effectively duplicating code that already exists in the library. So instead I propose a more general function in the library that handles conversion itself, keeping Repose cleaner.

**Comment by dmnjohns** [comment]
Closing in favor of #81.
